base-containers/base/inginious_container_api/run_student.py

```diff
@@ -99,17 +105,29 @@ def run_student(cmd, container=None,
 
         # send the fds and the command/workdir
         connection.sendmsg([b'S'], [(socket.SOL_SOCKET, socket.SCM_RIGHTS, array.array("i", [stdin, stdout, stderr]))])
-        connection.send(msgpack.dumps({"command": cmd, "working_dir": working_dir}))
+        connection.send(msgpack.dumps({"command": cmd, "working_dir": working_dir, "ssh": ssh, "user": user}))
 
         # Allow to send signals
         if signal_handler_callback is not None:
             def receive_signal(signum_s):
                 signum_data = str(signum_s).zfill(3).encode("utf8")
                 connection.send(signum_data)
-
             signal_handler_callback(receive_signal)
 
+        if ssh:  # The student_container will send id and password for ssh connection, transfer it to the agent
+            ssh_id = msgpack.loads(connection.recv(240))  # The message is asserted to be 240 bytes before sending
```

Comment:
There should be a msgpack API to compute exactually at runtime the expected message size, rather than hardcoding it. This is typically the kind of stuff that the library should abstract.

Reply:
The exact number of bytes can be computed at the student_container side (it's actually a constant: 54 bytes).
But since the whole goal here is to communicate between these 2 sides (student_container and grading_container) I don't see the issue about hardcoding the number of bytes (since it's a consant)  :/

Reply:
It is rarely a good practice to hardcode that kind of thing unless there is an explicit standardization that ensures this will never change. While msgpack should ensure the size of the message is constant through different architectures by its design, you never know if the internal message representation won't change some day as it is not standardized.

These kind of changes actually already happend once when msgpack was upgraded to version 1.0. So you should really rely on a msgpack for this.

The typical way msgpack deal with this is through its [Unpacker class](https://github.com/msgpack/msgpack-python/tree/main#streaming-unpacking). And this is already the way it is implemented in INGInious : https://github.com/UCL-INGI/INGInious/blob/b60326b702022ffe41391e7b7b003b3b3bbb8080/base-containers/base/bin/_run_student_intern#L59

And, while it is commented "It's ugly", there is actually no many other manners to do this with msgpack. It is simply not designed to support non-msgpack messages in the same stream. A possible alternative, in this really specific case, is to pack a dummy message, compute its size at runtime, and this should be the expected value.

It is clear the actual way to solve this "It's ugly" thing is to ensure only msgpack messages are sent in the stream, and I'm not requiring it. I just want to avoid really bad surprises if something changes in msgpack implementation.

Reply:
Should be better now, computing the expected size at runtime instead of hardcoding it.